cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,361 @@
+            lineage with the new descendant is added to the queue, but with its
+            status marked as active, and further events will be sampled at the
+            time the lineage divides. Additionally, its fitness will be updated.
+            The descendant node is added to the tree object, with the edge
+            weight between the current node and the descendant representing the
+            lifespan of the descendant. In the case the descendant would live
+            past the end of the experiment (both birth and death times exceed
+            past the end of the experiment), then the lifespan is cut off at
+            the experiment time and a final observed sample is added to the
+            tree. In this case the lineage is marked as inactive as well.
+
+            Args:
+                unique_id: The unique ID number to be used to name a new node
+                    added to the tree
+                lineage: The current extant lineage to extend. Contains the ID
+                    of the internal node to attach the descendant to, the
+                    current birth scale parameter of the lineage, the current
+                    total lived time of the lineage, and the status of whether
+                    the lineage is still dividing
+
+            Returns:
+                The new unique ID to use for the next event, incremented only if
+                a new descendant is added to the tree
+            """
+            assert(lineage["active"])
+
+            unique_id = next(names)
+
+            birth_waiting_time = self.birth_waiting_distribution(lineage["birth_scale"])
+            death_waiting_time = self.death_waiting_distribution()
+            if birth_waiting_time <= 0 or death_waiting_time <= 0:
+                raise TreeSimulatorError(
+                    "0 or negative waiting time detected"
+                )
+
+            # If birth and death would happen after the total experiment time,
+            # just cut off the living branch length at the experiment time
+            if (
+                self.experiment_time
+                and lineage["total_time"] + birth_waiting_time
+                >= self.experiment_time
+                and lineage["total_time"] + death_waiting_time
+                >= self.experiment_time
+            ):
+                tree.add_node(unique_id)
+                tree.nodes[unique_id]["birth_scale"] = lineage["birth_scale"]
+                tree.add_edge(
+                    lineage["id"],
+                    unique_id,
+                    weight=self.experiment_time - lineage["total_time"],
+                )
+                tree.nodes[unique_id]["total_time"] = self.experiment_time
+
+                current_lineages.put(
+                    (
+                        self.experiment_time,
+                        unique_id,
+                        {
+                            "id": unique_id,
```

Comment:
Why do we have the `unique_id` twice in this data structure?

Reply:
This is to provide an additional field so that the PriorityQueue can break ties in total_time. Otherwise, if two lineages have the same total time, then it is unable to break ties. I also wanted to keep the total_time and unique_id information in the lineage dictionary itself too, for readability purposes. Let me know if you disagree!

Reply:
Ah, I see, this is a very good reason, I hadn't noticed. Thanks Richard!